### PR TITLE
Type hinting for discord.commands

### PR DESCRIPTION
**Pull request by Luc1412**
## Summary

This is a work in progress pr for fully type hint discord commands.

Due to a larger restructure, I had to create a new pr. 
Original Pr: #133 

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

**Comment by Lulalaby** [comment]
Mark it as draft please

File: discord/bot.py [hunk 1]
```diff
@@ -346,7 +346,11 @@ async def register_commands(self) -> None:
                     )
                     raise
 
-    async def process_application_commands(self, interaction: Interaction) -> None:
+    async def process_application_commands(
+            self,
+            interaction: Interaction,
+            application_context_cls: Optional[Type[ApplicationContext]] = None
```
Comment (BobDotCom) [inline discord/bot.py:352]: Looks like some code from #424 got thrown in here.

Comment (Lulalaby) [reply discord/bot.py:352]: Lol

Comment (Luc1412) [reply discord/bot.py]: idk how it even came into this branch, but it should be solved now

File: discord/commands/_types.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+from typing import Callable, TYPE_CHECKING, Union, Coroutine, Any, TypeVar
```
Comment (RPSMain) [inline discord/commands/_types.py:1]: maybe add the license here?

**Review by RPSMain — APPROVED** [review]

**Comment by Lulalaby** [comment]
@Luc1412 are you still working on that?

**Comment by Luc1412** [comment]
> @Luc1412 are you still working on that?

I already continued on this. Somehow this pr broke, because the changes aren't shown here. Unfortunately am somehow busy until mid January and can't guarantee that I'll finish it until the planed release.

**Comment by Lulalaby** [comment]
oke

**Comment by Lulalaby** [comment]
I would merge that into our typing branch for now

**Pull request merged**